### ISSUE TITLE
Invalid value: "eksattachments": must be a standard resource type or fully qualified

### ISSUE THREAD

**Issue by excieve** [isindir/eks-ebs-node-operator#17]
Container resource name validation is still failing after #16.

```
# * spec.containers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified
# * spec.containers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers
# * spec.containers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified
# * spec.containers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers
```

It appears that any extended resource should follow the fully qualified resource name scheme (e.g. `eks.ebsnodeoperator/attachments`):
https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources

**Comment by isindir** [comment]
@excieve sorry missed this message somehow.

Could you please send me a bit more information about runtime environment:
* how k8s was installed (eksctl, kops, kubespray, other) and which version of the k8s and or installer you are using ?
* do you use any policy management system which may block nonstandard fields ?
* can you please share example pod definition ?
* can you please share events section from pod with limitation added from command `kubectl describe po` ?

Thanks

**Comment by excieve** [comment]
@isindir No worries, instead of using EBS volumes we ended up deploying Longhorn, which is more suitable for our purposes anyway.

To answer your questions:
 - This was tested on EKS 1.14.x deployed using Terraform and Rancher.
 - Just PSPs, but I don't think they block anything like this.

Unfortunately, I don't have examples and events anymore, sorry. I believe the problem was k8s API server not accepting a spec at all because of extended resource name not being an accepted FQDN.

**Comment by lagivan** [comment]
@isindir facing the same issue now. So I'll answer your questions:

- how k8s was installed (eksctl, kops, kubespray, other) and which version of the k8s and or installer you are using ?
Amazon EKS, kubernetes version 1.15, deployed with Terraform
- do you use any policy management system which may block nonstandard fields ?
No
- can you please share example pod definition ?
```
apiVersion: apps/v1
kind: StatefulSet
metadata:
  annotations:
  labels:
    app: tracemaster
  name: tracemaster
  namespace: prod
spec:
  replicas: 1
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: tracemaster
  serviceName: tracemaster
  template:
    metadata:
      labels:
        app: tracemaster
    spec:
      nodeSelector:
        nodetype: tracemaster
      tolerations:
        - key: dedicated
          operator: Equal
          value: tracemaster
      initContainers:
      - name: init-data
        image: X.dkr.ecr.eu-west-1.amazonaws.com/aws-cli:latest
        imagePullPolicy: IfNotPresent
        command:
        - aws
        - s3
        - sync
        - s3://X
        - /aws
        - --delete
        - --no-progress
        volumeMounts:
         - mountPath: /aws
           name: my-data
        resources:
          limits:
            eksattachments: "1"
          requests:
            eksattachments: "1"
        env:
        - name: AWS_REGION
          value: eu-west-1
        - name: AWS_ACCESS_KEY_ID
          valueFrom:
            secretKeyRef:
              name: job-user
              key: AWS_ACCESS_KEY_ID
        - name: AWS_SECRET_ACCESS_KEY
          valueFrom:
            secretKeyRef:
              name: job-user
              key: AWS_SECRET_ACCESS_KEY
      containers:
      - name: tracemaster
        image: X.dkr.ecr.eu-west-1.amazonaws.com/tracemaster:1.1.2
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 80
          protocol: TCP
        volumeMounts:
        - mountPath: /home/tracemaster/data
          name: my-data
        resources:
          limits:
            cpu: "1.5"
            eksattachments: "1"
          requests:
            memory: "12Gi"
            cpu: "1.0"
            eksattachments: "1"
      volumes:
        - name: my-data
          persistentVolumeClaim:
            claimName: my-data-claim
```
 
- can you please share events section from pod with limitation added from command kubectl describe po ?
Events:
```
  Type     Reason        Age                   From                    Message
  ----     ------        ----                  ----                    -------
  Warning  FailedCreate  2m59s (x18 over 13m)  statefulset-controller  create Pod tracemaster-0 in StatefulSet tracemaster failed error: Pod "tracemaster-0" is invalid: [spec.containers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified, spec.containers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers, spec.containers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified, spec.containers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers, spec.initContainers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified, spec.initContainers[0].resources.limits[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers, spec.initContainers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource type or fully qualified, spec.initContainers[0].resources.requests[eksattachments]: Invalid value: "eksattachments": must be a standard resource for containers]
```

**Comment by lagivan** [comment]
As suggested by @excieve, I've tried to replace `eksattachments` by `eks.ebsnodeoperator/attachments` in `eks-ebs-node-operator/pkg/controller/node/node_controller.go`. I've built a new docker image and used it instead. However, it fails to reconcile nodes with the following log from `eks-ebs-node-operator`:

```
2021-01-19T14:03:45.390093272Z {"level":"info","ts":1611065025.3888657,"logger":"controller_node","msg":"Reconciling Node","Request.Name":"ip-10-141-1-148.eu-west-1.compute.internal"}
2021-01-19T14:03:45.390148883Z {"level":"info","ts":1611065025.388925,"logger":"controller_node","msg":"Reconcile: node capacity must be set to","Request.Name":"ip-10-141-1-148.eu-west-1.compute.internal","NodeType":"t3a.large","MaxVolumes":"24"}
2021-01-19T14:03:45.395673915Z {"level":"info","ts":1611065025.3955028,"logger":"controller_node","msg":"Failed to patch the node","Request.Name":"ip-10-141-1-148.eu-west-1.compute.internal","NodeName":"ip-10-141-1-148.eu-west-1.compute.internal"}
2021-01-19T14:03:45.395732245Z {"level":"error","ts":1611065025.3955443,"logger":"controller-runtime.controller","msg":"Reconciler error","controller":"node-controller","request":"/ip-10-141-1-148.eu-west-1.compute.internal","error":"the server rejected our request due to an error in our request","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/user/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/user/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:258\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/user/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/user/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/user/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/user/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/user/go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88"}
```

So I don't have a working solution yet. Any help is appreciated.

**Comment by lagivan** [comment]
Ok, I've made it work. The problem was that / must be encoded as ~1 in the patch path (mentioned here - https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/).

So I've changed the line in `eks-ebs-node-operator/pkg/controller/node/node_controller.go` to:

`const extendedResourceName string = "eks.ebsnodeoperator~1attachments"`

Then `eks.ebsnodeoperator/attachments` should be set in resource limits/requests for pods with mounted EBS volumes.

@isindir would be nice to have this fix incorporated.
